### PR TITLE
[Neo Core StdLib] Add Base64url

### PR DESCRIPTION
**Pull request by Jim8y**
# Description

This pr answers https://github.com/neo-project/neo/issues/3451 and adds base64url support.

Fixes # https://github.com/neo-project/neo/issues/3451

## Type of change

- [ ] Optimization (the change is only an optimization)
- [ ] Style (the change is only a code style for better maintenance or standard purpose)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] TestBase64Url

**Test Configuration**:

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/Neo/SmartContract/Native/StdLib.cs [hunk 1]
```diff
@@ -169,12 +169,41 @@ public static string Base58CheckEncode([MaxLength(MaxInputLength)] byte[] data)
         /// </summary>
         /// <param name="s">The base58 <see cref="string"/>.</param>
         /// <returns>The decoded byte array.</returns>
-        [ContractMethod(CpuFee = 1 << 16)]
+        [ContractMethod(Hardfork.HF_Echidna, CpuFee = 1 << 16)]
         public static byte[] Base58CheckDecode([MaxLength(MaxInputLength)] string s)
```
Comment (AnnaShaleva) [inline src/Neo/SmartContract/Native/StdLib.cs:173]: Is it a typo? I suppose we should keep `Base58CheckDecode` method unchanged.

File: src/Neo/SmartContract/Native/StdLib.cs [hunk 2]
```diff
@@ -169,12 +169,41 @@ public static string Base58CheckEncode([MaxLength(MaxInputLength)] byte[] data)
...
+            return base64.TrimEnd('=').Replace('+', '-').Replace('/', '_');
+        }
+
+        /// <summary>
+        /// Decodes a byte array from a base64Url string.
+        /// </summary>
+        /// <param name="s">The base64Url string.</param>
+        /// <returns>The decoded byte array.</returns>
+        [ContractMethod(CpuFee = 1 << 5)]
+        public static byte[] Base64UrlDecode([MaxLength(MaxInputLength)] string s)
```
Comment (AnnaShaleva) [inline src/Neo/SmartContract/Native/StdLib.cs:196]: This method should be active starting from Echidna.

File: src/Neo/SmartContract/Native/StdLib.cs (hunk 1 above)
Comment (Jim8y) [reply src/Neo/SmartContract/Native/StdLib.cs:173]: my fault,,,,,,, wrong place.

File: src/Neo/SmartContract/Native/StdLib.cs [hunk 3]
```diff
@@ -131,6 +131,35 @@ public static byte[] Base64Decode([MaxLength(MaxInputLength)] string s)
...
+        /// <returns>The decoded byte array.</returns>
+        [ContractMethod(Hardfork.HF_Echidna, CpuFee = 1 << 5)]
+        public static byte[] Base64UrlDecode([MaxLength(MaxInputLength)] string s)
+        {
+            var padded = s.Replace('-', '+').Replace('_', '/');
+            switch (padded.Length % 4)
+            {
+                case 2: padded += "=="; break;
+                case 3: padded += "="; break;
+            }
```
Comment (cschuchardt88) [inline src/Neo/SmartContract/Native/StdLib.cs:159]: What is this for? This is wrong.

 - Padding is skipped so the pad character '=' doesn't have to be percent encoded.
 - The 62nd and 63rd regular base64 encoding characters ('+' and '/') are replaced with ('-' and '_'). This makes the encoding alphabet URL safe.

**Review by cschuchardt88 — CHANGES_REQUESTED** [review]

File: src/Neo/SmartContract/Native/StdLib.cs (hunk 3 above)
Comment (Jim8y) [reply src/Neo/SmartContract/Native/StdLib.cs]: updated

**Review by cschuchardt88 — APPROVED** [review]

**Review by shargon — APPROVED** [review]

File: src/Neo/SmartContract/Native/StdLib.cs [hunk 4]
```diff
@@ -131,6 +132,28 @@ public static byte[] Base64Decode([MaxLength(MaxInputLength)] string s)
...
+            return Base64UrlEncoder.Encode(data);
+        }
+
+        /// <summary>
+        /// Decodes a byte array from a base64Url string.
+        /// </summary>
+        /// <param name="s">The base64Url string.</param>
+        /// <returns>The decoded base64Url string.</returns>
+        [ContractMethod(Hardfork.HF_Echidna, CpuFee = 1 << 5)]
+        public static string Base64UrlDecode([MaxLength(MaxInputLength)] string s)
```
Comment (Wi1l-B0t) [inline src/Neo/SmartContract/Native/StdLib.cs:152]: Why the return type of `Base64UrlDecode` is `string`, but `Base64Decode` returns `byte[]`?

File: src/Neo/SmartContract/Native/StdLib.cs [hunk 5]
```diff
@@ -131,6 +132,28 @@ public static byte[] Base64Decode([MaxLength(MaxInputLength)] string s)
             return Convert.FromBase64String(s);
         }
 
+        /// <summary>
+        /// Encodes a byte array into a base64Url string.
+        /// </summary>
+        /// <param name="data">The base64Url to be encoded.</param>
+        /// <returns>The encoded base64Url string.</returns>
+        [ContractMethod(Hardfork.HF_Echidna, CpuFee = 1 << 5)]
+        public static string Base64UrlEncode([MaxLength(MaxInputLength)] string data)
```
Comment (Wi1l-B0t) [inline src/Neo/SmartContract/Native/StdLib.cs:141]: Why the input type of `Base64UrlEncode` is `string`, but the input type of `Base64UrlEncode` is `byte[]`?

**Pull request merged**